### PR TITLE
feat: Given a day of the voting date find the hour that maximizes the probability that the upgrade happens during working hours

### PR DESCRIPTION
**Pull request by stedfn**
Usage:

`python3 estimate_epoch_start_time.py --chain_id mainnet --voting_date_day "2024-10-23" --timezone cet`

<details><summary>Example Output</summary>

```
Epoch -1: 13 hours, 19 minutes
Epoch -2: 13 hours, 15 minutes
Epoch -3: 13 hours, 23 minutes
Epoch -4: 13 hours, 17 minutes

Exponential weighted average epoch length: 13 hours, 18 minutes
Predicted start of epoch 1: 2024-10-22 05:18:32 CEST+0200 Tuesday
Predicted start of epoch 2: 2024-10-22 18:37:26 CEST+0200 Tuesday
Predicted start of epoch 3: 2024-10-23 07:56:19 CEST+0200 Wednesday
Predicted start of epoch 4: 2024-10-23 21:15:12 CEST+0200 Wednesday
Predicted start of epoch 5: 2024-10-24 10:34:05 CEST+0200 Thursday
Predicted start of epoch 6: 2024-10-24 23:52:58 CEST+0200 Thursday
Predicted start of epoch 7: 2024-10-25 13:11:51 CEST+0200 Friday
Predicted start of epoch 8: 2024-10-26 02:30:45 CEST+0200 Saturday
Predicted start of epoch 9: 2024-10-26 15:49:38 CEST+0200 Saturday
Predicted start of epoch 10: 2024-10-27 04:08:31 CET+0100 Sunday

Voting hours on 2024-10-23 UTC that result in upgrade during working hours (UTC 8:00-22:00):
- 00:00, Upgrade Epoch: 4
- 01:00, Upgrade Epoch: 4
- 02:00, Upgrade Epoch: 4
- 03:00, Upgrade Epoch: 4
- 04:00, Upgrade Epoch: 4
- 05:00, Upgrade Epoch: 4
- 06:00, Upgrade Epoch: 5
- 07:00, Upgrade Epoch: 5
- 08:00, Upgrade Epoch: 5
- 09:00, Upgrade Epoch: 5
- 10:00, Upgrade Epoch: 5
- 11:00, Upgrade Epoch: 5
- 12:00, Upgrade Epoch: 5
- 13:00, Upgrade Epoch: 5
- 14:00, Upgrade Epoch: 5
- 15:00, Upgrade Epoch: 5
- 16:00, Upgrade Epoch: 5
- 17:00, Upgrade Epoch: 5
- 18:00, Upgrade Epoch: 5
- 19:00, Upgrade Epoch: 5
``` 

</details>

### CODE REVIEW

File: debug_scripts/estimate_epoch_start_time.py [hunk 1]
```diff
@@ -122,11 +122,55 @@ def find_epoch_for_timestamp(future_epochs, voting_datetime):
     return len(future_epochs)
 
 
+def find_best_voting_hour(voting_date_str, future_epochs, target_timezone):
+    WORKING_HOURS_START = 8  # 8:00 UTC
```
[the remark this answers is not in the document]
Comment (stedfn) [reply debug_scripts/estimate_epoch_start_time.py:123]: If I am not missing something I think it's fine if I localize the voting date to the target timezone and then use the unix timestamp

File: debug_scripts/estimate_epoch_start_time.py [hunk 2]
```diff
@@ -119,6 +119,54 @@ def find_epoch_for_timestamp(future_epochs, voting_timestamp):
...
+            print(
+                "Not enough future epochs predicted to determine all the protocol upgrade times."
+            )
+            break
+
+        protocol_upgrade_timestamp = future_epochs[protocol_upgrade_epoch_number
+                                                   - 1]
+        protocol_upgrade_datetime = datetime.fromtimestamp(
+            protocol_upgrade_timestamp)
+        upgrade_datetime_utc = protocol_upgrade_datetime.astimezone(pytz.utc)
```
Comment (VanBarbascu) [inline debug_scripts/estimate_epoch_start_time.py:152]: Why do we need to convert this datetime to UTC? In `predict_future_epochs`, we kept all `future_epochs` in UTC.

File: debug_scripts/estimate_epoch_start_time.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (VanBarbascu) [reply debug_scripts/estimate_epoch_start_time.py:123]: It should work.

File: debug_scripts/estimate_epoch_start_time.py (hunk 2 above)
Comment (stedfn) [reply debug_scripts/estimate_epoch_start_time.py:152]: you are right, I forgot it from the previous version where future_epochs was converted to the target timezone

File: debug_scripts/estimate_epoch_start_time.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (stedfn) [reply debug_scripts/estimate_epoch_start_time.py:123]: My understanding is that the `timestamp()` function will give the timestamp in UTC

[the remark this answers is not in the document]
Comment (VanBarbascu) [reply debug_scripts/estimate_epoch_start_time.py:123]: That is correct.

File: debug_scripts/estimate_epoch_start_time.py [hunk 3]
```diff
@@ -119,6 +119,53 @@ def find_epoch_for_timestamp(future_epochs, voting_timestamp):
...
+            )
+            break
+
+        protocol_upgrade_timestamp = future_epochs[protocol_upgrade_epoch_number
+                                                   - 1]
+        protocol_upgrade_datetime = datetime.fromtimestamp(
+            protocol_upgrade_timestamp)
+        upgrade_hour_utc = protocol_upgrade_datetime.hour
+
+        if WORKING_HOURS_START <= upgrade_hour_utc < WORKING_HOURS_END:
```
Comment (VanBarbascu) [inline debug_scripts/estimate_epoch_start_time.py:154]: We should use `target_timezone` to find the voting time that will rezult in upgrade happening during the working hours of said `target_timezone`. For this, we should localize `protocol_upgrade_datetime`.

File: debug_scripts/estimate_epoch_start_time.py [hunk 4]
```diff
@@ -119,6 +119,53 @@ def find_epoch_for_timestamp(future_epochs, voting_timestamp):
...
+            valid_hours.append((hour, protocol_upgrade_epoch_number))
+
+    if valid_hours:
+        print(
+            f"\nVoting hours on {voting_date_str} {target_timezone} that result in upgrade during working hours (UTC {WORKING_HOURS_START}:00-{WORKING_HOURS_END}:00):"
```
Comment (VanBarbascu) [inline debug_scripts/estimate_epoch_start_time.py:159]: When doing a release, we set the voting date in the `UTC`. It is useful to have this output in the following format: voting date in UTC -> upgrade date in working hours for `target_timezone`.

File: debug_scripts/estimate_epoch_start_time.py (hunk 3 above)
Comment (stedfn) [reply debug_scripts/estimate_epoch_start_time.py:154]: I see what you mean now but my understanding has been that we want the upgrade to happen while we have engineers online (any), which would be between 8 - 22 UTC. Maybe @jancionear can you please clarify or provide more details for this feature?

File: debug_scripts/estimate_epoch_start_time.py (hunk 4 above)
Comment (stedfn) [reply debug_scripts/estimate_epoch_start_time.py]: Is it good if I display the upgrade epoch and before that the epochs are listed in the `target_timezone` (as in the example output)?

**Review by VanBarbascu — APPROVED** [review]
Looks great! Thanks for taking the time to work on this!

**Pull request merged**